### DISCUSSION TITLE
Write with response

### DISCUSSION THREAD

**Discussion by HaniFakhouri** [JuulLabs/kable#507] [Q&A] [answered]
Hi. If I write to characteristic using writeType set to WriteType.WithResponse, how can I observe/consume the response?

**Comment by twyatt** [comment, marked as answer]
Write with response configures whether or not the link layer will handle the acknowledgement of receipt of a packet. Internally in Kable, we only have visibility into _when_ the link layer is ready to receive the next packet. So, write with response has little to do with getting data back from your peripheral.

How you get data from your peripheral is really up to you, but a common design I've seen is to have your peripheral have a characteristic dedicated to writes (we'll call it `A`) and a characteristic for observing "responses" (characteristic changes — we'll call it `B`):

```kotlin
val response = B.onSubscription { A.write(request) }.first()
```

Some considerations with the above pseudo code:
- You'll want to make sure `B` continually has subscribers, otherwise you'll spin up/down the observation for every write (this can be done with an `observe(..).shareIn(..)` with `Eagerly`, or by means of having another subscriber of `B` remain active
- You may want to have a monotonic ID in your `request` that your peripheral sends back with the "response" so that you can properly associate the response with the request (e.g. `first { response.id == request.id }`)
